File: typescript/packages/bench/src/reports/technicianLoadReport.ts
/** Booked share of each technician's day. */

import type { ReportRow } from "../contracts/reportContract";
import type { Dataset } from "../dataset";
import { reportRow } from "../contracts/reportContract";
import { AbstractReport } from "../support/abstractReport";

/** Booked share of each technician's day. */
export class TechnicianLoadReport extends AbstractReport {
    /** Registry slug; also the URL segment. */
    static readonly SLUG = "technician-load";

    constructor() {
        super(TechnicianLoadReport.SLUG, "Technician load");
    }

    override rows(data: Dataset): ReportRow[] {
        return data.technicians.map((technician) => reportRow(technician.name, technician.utilisation()));
    }
}

File: rust/crates/atelier-bench/src/reports/technician_load_report.rs
//! Booked share of each technician's day.

use crate::contracts::report_contract::{ReportContract, ReportRow};
use crate::dataset::Dataset;
use crate::support::abstract_report::ReportBase;

/// Booked share of each technician's day.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TechnicianLoadReport {
    base: ReportBase,
}

impl TechnicianLoadReport {
    /// Registry slug; also the URL segment.
    pub const SLUG: &'static str = "technician-load";

    pub const fn new() -> Self {
        Self {
            base: ReportBase::new(Self::SLUG, "Technician load"),
        }
    }

    /// Shared identity and rendering precision.
    pub const fn base(&self) -> &ReportBase {
        &self.base
    }
}

impl Default for TechnicianLoadReport {
    fn default() -> Self {
        Self::new()
    }
}

impl ReportContract for TechnicianLoadReport {
    fn slug(&self) -> &'static str {
        self.base.slug()
    }

    fn title(&self) -> &'static str {
        self.base.title()
    }

    fn rows(&self, data: &Dataset) -> Vec<ReportRow> {
        data.technicians
            .iter()
            .map(|technician| ReportRow::new(technician.name.clone(), technician.utilisation()))
            .collect()
    }
}
